lib/events/filesessions/fileasync.go

```diff
@@ -375,6 +378,59 @@ func (u *upload) removeFiles() error {
 	return trace.NewAggregate(errs...)
 }
 
+var errNotEncrypted = errors.New("recording is not encrypted")
+
+func (u *Uploader) uploadEncryptedRecording(ctx context.Context, sessionID string, in io.ReadCloser) error {
+	partIter := func(yield func([]byte, error) bool) {
+		var buf bytes.Buffer
+		for {
+			buf.Reset()
+			header, err := events.ParsePartHeader(in)
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+
+				yield(nil, trace.Wrap(err))
+				return
+			}
+
+			if header.Flags&events.ProtoStreamFlagEncrypted == 0 {
+				yield(nil, trace.Wrap(errNotEncrypted))
+				return
+			}
+
+			if _, err := buf.Write(header.Bytes()); err != nil {
+				yield(nil, trace.Wrap(err))
+				return
+			}
+
+			totalPartSize := int64(header.PartSize + header.PaddingSize)
+			reader := io.LimitReader(in, totalPartSize)
+			copied, err := io.Copy(&buf, reader)
+			if err != nil && !errors.Is(err, io.EOF) {
```

Comment:
shouldn't we break when we get an EOF? Or is it expected to still yield and break at the next ParsePartHeader?

Reply:
It should yield and break at the next `ParsePartheader`, yeah. This EOF would come from reaching the limit of the `io.LimitReader` which doesn't necessarily mean we're done reading the file, just the current part.